File: corestream-node/corestream.js
const net = require('net');
const protobuf = require('protobufjs');

class CoreStreamClient {
    constructor(host, port, apiKey) {
        this.host = host;
        this.port = port;
        this.apiKey = apiKey;
        this.client = new net.Socket();
        this.root = null;
    }

    async connect() {
        // Load the Protobuf schema dynamically
        this.root = await protobuf.load('../proto/messages.proto');
        
        return new Promise((resolve, reject) => {
            this.client.connect(this.port, this.host, () => {
                const AuthHandshake = this.root.lookupType('AuthHandshake');
                const errMsg = AuthHandshake.verify({ api_key: this.apiKey });
                if (errMsg) throw Error(errMsg);

                const message = AuthHandshake.create({ api_key: this.apiKey });
                const buffer = AuthHandshake.encode(message).finish();

                const lengthBuf = Buffer.alloc(4);
                lengthBuf.writeUInt32BE(buffer.length + 1, 0);

                this.client.write(lengthBuf);
                this.client.write(Buffer.from([4])); // MSG_TYPE_AUTH
                this.client.write(buffer);
            });

            this.client.once('data', (data) => {
                if (data[0] === 1) {
                    console.log(`✅ Connected and Authenticated to CoreStream at ${this.host}:${this.port}`);
                    resolve();
                } else {
                    reject(new Error("CoreStream Authentication Failed. Invalid API Key."));
                }
            });

            this.client.on('error', reject);
        });
    }

    async produce(topic, dataString) {
        const ProducerPayload = this.root.lookupType('ProducerPayload');
        
        const payload = {
            topic: topic,
            data: Buffer.from(dataString),
            timestamp: Math.floor(Date.now() / 1000)
        };

        const errMsg = ProducerPayload.verify(payload);
        if (errMsg) throw Error(errMsg);

        const message = ProducerPayload.create(payload);
        const buffer = ProducerPayload.encode(message).finish();

        const lengthBuf = Buffer.alloc(4);
        lengthBuf.writeUInt32BE(buffer.length + 1, 0);

        this.client.write(lengthBuf);
        this.client.write(Buffer.from([0])); // MSG_TYPE_PRODUCER
        this.client.write(buffer);
    }
}

// Example Usage
(async () => {
    try {
        const client = new CoreStreamClient('127.0.0.1', 9092, 'super_secret_corestream_key');
        await client.connect();
        await client.produce('node_logs', 'Hello from the official Node.js SDK!');
        console.log('🚀 Successfully published message via Node SDK!');
        process.exit(0);
    } catch (err) {
        console.error(err);
        process.exit(1);
    }
})();

File: corestream-python/corestream.py
import socket
import struct
import time
import messages_pb2

class CoreStreamClient:
    def __init__(self, host: str, port: int, api_key: str):
        self.host = host
        self.port = port
        self.api_key = api_key
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.connect()

    def connect(self):
        self.socket.connect((self.host, self.port))
        
        # 1. Send AuthHandshake
        auth = messages_pb2.AuthHandshake(api_key=self.api_key)
        auth_bytes = auth.SerializeToString()
        
        # Length is payload length + 1 byte for MSG_TYPE
        length = len(auth_bytes) + 1
        self.socket.sendall(struct.pack(">I", length))
        
        # MSG_TYPE_AUTH = 4
        self.socket.sendall(bytes([4]))
        self.socket.sendall(auth_bytes)
        
        # Await ACK
        ack = self.socket.recv(1)
        if not ack or ack[0] != 1:
            raise PermissionError("CoreStream Authentication Failed. Invalid API Key.")
            
        print(f"✅ Connected and Authenticated to CoreStream at {self.host}:{self.port}")

    def produce(self, topic: str, data: bytes):
        payload = messages_pb2.ProducerPayload(
            topic=topic,
            data=data,
            timestamp=int(time.time())
        )
        payload_bytes = payload.SerializeToString()
        
        length = len(payload_bytes) + 1
        self.socket.sendall(struct.pack(">I", length))
        
        # MSG_TYPE_PRODUCER = 0
        self.socket.sendall(bytes([0]))
        self.socket.sendall(payload_bytes)

# Example Usage
if __name__ == "__main__":
    client = CoreStreamClient("127.0.0.1", 9092, "super_secret_corestream_key")
    client.produce("python_logs", b"Hello from the official Python SDK!")
    print("🚀 Successfully published message via Python SDK!")
